mkosi

```diff
@@ -2204,42 +2259,125 @@ def install_arch(args: CommandLineArguments, root: str, do_run_build_script: boo
-            "--config", pacman_conf,
-        ]
+        os.makedirs(hooks_dir, 0o755, exist_ok=True)
+        os.makedirs(scripts_dir, 0o755, exist_ok=True)
 
-        try:
-            return run(cmdline + args, check=True)
-        finally:
-            stop_gpg_agent()
+        # Disable depmod pacman hook as depmod is handled by kernel-install as well.
+        os.symlink("/dev/null", os.path.join(hooks_dir, "60-depmod.hook"))
+
+        kernel_add_hook = os.path.join(hooks_dir, "mkosi-kernel-add.hook")
+        with open(kernel_add_hook, 'w') as f:
+            f.write(
+                dedent(
+                    """\
+                    [Trigger]
+                    Operation = Install
+                    Operation = Upgrade
+                    Type = Path
+                    Target = usr/lib/modules/*/vmlinuz
+
+                    [Trigger]
+                    Operation = Install
+                    Operation = Upgrade
+                    Operation = Remove
+                    Type = Package
+                    Target = systemd
+                    Target = dracut
+                    Target = intel-ucode
+                    Target = amd-ucode
+
+                    [Action]
+                    Description = Adding kernel and initramfs images to /boot...
+                    When = PostTransaction
+                    Exec = /etc/pacman.d/scripts/mkosi-kernel-add
+                    NeedsTargets
+                    """
+                )
+            )
+
+        kernel_add_script = os.path.join(scripts_dir, "mkosi-kernel-add")
+        with open(kernel_add_script, 'w') as f:
+            f.write(
+                dedent(
+                    """\
+                    #!/bin/bash -e
+
+                    declare -a kernel_version
+
+                    # Check the targets passed by the pacman hook.
+                    while read -r line
+                    do
+                        if [[ "$line" =~ usr/lib/modules/([^/]+)/vmlinuz ]]
+                        then
+                            kernel_version+=( "${BASH_REMATCH[1]}" )
+                        else
+                            # If a non-matching line is passed, just rebuild all kernels.
+                            kernel_version=()
+                            for f in /usr/lib/modules/*/vmlinuz
+                            do
+                                kernel_version+=( "$(basename "$(dirname "$f")")" )
+                            done
+                            break
+                        fi
+                    done
+
+                    # (re)build the kernel images.
+                    for kv in "${kernel_version[@]}"
+                    do
+                        kernel-install add "$kv" "/usr/lib/modules/${kv}/vmlinuz"
+                    done
+                    """
+                )
+            )
+
+        make_executable(kernel_add_script)
+
+        kernel_remove_hook = os.path.join(hooks_dir, "mkosi-kernel-remove.hook")
```

Comment:
This hook should also get a numerical prefix, e.g. `60-`, since this is the number of the appropriate mkinitcpio hook.